### DISCUSSION TITLE
Detailed statistics for cache

### DISCUSSION THREAD

**Discussion by ferferga** [AdguardTeam/AdGuardHome#4063] [Ideas]
Hello!

I was a long-time PiHole user and, so far, I'm really happy with adguard (so much hassle-free than pi-hole, specially regarding all the dependencies and the possibility to properly run it with systemd, so kudos to all the devs!). However, there's one thing I find AdGuard very lackluster: **detailed cache usage stats**. Some comparisons vs Pi-Hole:

* In Pi-Hole, you can check at a glance (in a circular diagram) where each request goes: if to any of the upstream servers or to the PiHole's internal cache:
![image](https://user-images.githubusercontent.com/10274099/147886240-9557ae1e-b3f8-4e6a-9e05-c0dfdc2f7e84.png)
*Sample image from their web*

   * This graph would be useful, not only for cache, but also to know which of the upstream resolvers replies the most, how fast it does it and how well does the algorithm (balancing, parallel, fastest) works better for the user, allowing to take users informed decisions.

* In Pi-Hole you set the cache size specifying the max number of queries you want to have cached. You also can check the number of queries cached and how many evictions happened, so you can act on those stats and increase/decrease the cache.
I personally prefer AdGuard's approach to this (setting a max size in bytes), so I know exactly what to expect in regards to memory usage, however, I think giving the option to choose between both ways of setting cache size is a net benefit for everybody.

   * Additionally, I think the option to set the size should be given in bytes and kilobytes, so it's easier for *non-technically-inclined* people to do the math.

* There's no way to check exactly which queries are cached. I could use query logging, but if my query logging is for 6 hours, I'm losing a lot of information.
* There's no way to check how much of that cache is being used

Mikrotik routers have, in my opinion, the best fine-control of this:

**IP > DNS**
![image](https://user-images.githubusercontent.com/10274099/147886558-b8ae99ac-947d-4b67-93a1-86c8bf41723f.png)

**IP > DNS > Cache**
*I collapsed the columns for privacy*
![image](https://user-images.githubusercontent.com/10274099/147886584-830bbc8a-1d75-41db-8f0f-1fcda575d368.png)

I think all of this will please a lot of hardcore users like me, but will also make easier to understand what cache is, how it works and how it speeds your navigation to the *non-technically-inclined* users as well!

**Comment by ainar-g** [comment]
Hello!  The cache statistics is feature request #1453.  Per-upstream statistics is #2013 and some others.  I'm not quite sure what you mean by “which queries are cached”, since the cache should cover all queries, so logically the domains that are requested the most are also cached the most.

**Reply by ferferga** [reply]
Hello @ainar-g! Searched discussions and issues but couldn't find anything related to my request (sorry for the troubles!)

Let me clarify the "which queries are cached part": Take a look at Mikrotik's DNS query list, there's a TTL for each request. As you say, all the queries are covered by cache but, if I'm not using optimistic cache, the TTL will remove some of them from cache. Right now, I don't know which queries have passed that TTL and are, thus, removed from cache.

This is also useful in the situations when the user has set the cache size to a low value or the server gets a lot of queries, so you can check which old queries are removed (even when their TTL hasn't passed) due to the cache exceeding the max size.

**Reply by ainar-g** [reply]
I'm not entirely sure if per-host statistics on cache hits and misses is something we really want, to be honest.  But we'll see.  It's much too early right now to know for sure, since we don't even have a storage format for the future statistics finalized yet.
